File: 0931-minimum-falling-path-sum/0931-minimum-falling-path-sum.cpp

class Solution {
public:
    int minFallingPathSum(vector<vector<int>>& A) {
        int m = A.size();
        int n = A[0].size();

        if (m == 1 || n == 1) return A[0][0];

        vector<vector<int>> dp(m,vector<int>(n, INT_MAX));
        int ans = INT_MAX;

        for (int i = 0; i < A.size(); ++i) {
            ans = min(ans, minFallingPathSum(A, 0, i, dp));
        }

        return ans;
    }

    int minFallingPathSum(vector<vector<int>>& A, int row, int col,  vector<vector<int>>& dp) {
        int m = A.size();
        int n = A[0].size();

        if (dp[row][col] != INT_MAX) return dp[row][col];

        if (row == m - 1)
            return dp[row][col] = A[row][col];

        int left = INT_MAX, right = INT_MAX;

        if (col > 0)
            left = minFallingPathSum(A, row + 1, col - 1, dp);

        int straight = minFallingPathSum(A, row + 1, col, dp);

        if (col < n - 1)
            right = minFallingPathSum(A, row + 1, col + 1, dp);

        dp[row][col] = min(left, min(straight, right)) + A[row][col];

        return dp[row][col];
    }
};

File: 0931-minimum-falling-path-sum/0931-minimum-falling-path-sum.java
class Solution {
    public int minFallingPathSum(int[][] matrix) {
        // int res = Integer.MAX_VALUE;
        // for (int j = 0; j < matrix[0].length; j++) {
        // res = Math.min(res, solve(matrix, matrix.length - 1, j));
        // }
        // return res;
        return tabulation(matrix);
    }

    public int solve(int[][] matrix, int i, int j) {
        if (j < 0 || j >= matrix[0].length) {
            return Integer.MAX_VALUE;
        }
        if (i == 0) {
            return matrix[i][j];
        }
        int st = solve(matrix, i - 1, j);
        int left = solve(matrix, i - 1, j - 1);
        int right = solve(matrix, i - 1, j + 1);
        return matrix[i][j] + Math.min(Math.min(st, left), right);
    }

    public int tabulation(int[][] matrix) {
        int[][] dp = new int[matrix.length][matrix[0].length];
        int st, left, right = Integer.MAX_VALUE;
        for (int j = 0; j < matrix[0].length; j++) {
            dp[0][j] = matrix[0][j];
        }
        for (int i = 1; i < matrix.length; i++) {
            for (int j = 0; j < matrix[0].length; j++) {
                st = dp[i - 1][j];
                left = j > 0 ? dp[i - 1][j - 1] : Integer.MAX_VALUE;
                right = j < matrix[0].length - 1 ? dp[i - 1][j + 1] : Integer.MAX_VALUE;
                dp[i][j] = matrix[i][j] + Math.min(Math.min(st, left), right);
            }
        }
        int res = Integer.MAX_VALUE;
        for (int j = 0; j < matrix[0].length; j++) {
            res = Math.min(res, dp[matrix.length - 1][j]);
        }
        return res;
    }
}